RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md

```diff
@@ -0,0 +1,323 @@
+* **Lowering with `torch.fx`** We can lower directly with `torch.fx` transformations. Please take a look at Extending PyTorch Quantization to Custom Backends for an example implementation of lowering in fx, for this we need to make sure all backend operators are exposed in `torch` namespace, for example: `torch.ops.custom_backend.quantized_conv2d`
+* **Lowering with TorchScript** We can also first script/trace the model and then lower with TorchScript, we need to make sure TorchScript pass can recognize the reference quantized functions, for example, this can be done by looking at the name of a CallFunction Node
+* **Lowering with a combination of 1 and 2** People can also do some transformations in `torch.fx` and then script/trace the model and do some other transformations in TorchScript, for example, in `torch.fx` we can transform reference quantized functions to actual (dequant - float_op - quant) pattern and in TorchScript we can fuse the patterns
+* **Custom Lowering** Users can also provide their own lowering functions that does not use `torch.fx` or TorchScript that can transform a Reference Quantized Model to a model that’s runnable on the target backend. Basically Reference Quantized Model is the standard format that is expected by backend developers.
+
+## Use Case 3: Extending Quantization to Support Custom Quantized Operators and Fusions for Inference in Custom Backend
+
+If a backend supports more quantized operators, more fusions, different ways of quantization, we will need to customize the flow. We only provide limited support for quantized (and fused) operator configurations at this point and we’ll work on improving the support in the future to allow arbitrary custom behaviors.
+
+Here we’ll give an example of supporting custom quantized operators and fusions.
+
+### Custom Operator support
+If you need to support an operator that does not have a reference operator implementation, you will need to implement a custom function/module to specify the quantized operator implementation. In addition, you will need to write a handler class that specifies how to convert a observed node to a quantized node.
+
+#### Example: Quantized BMM
+To add support for new quantized ops, we need to do the following:
+* Get fbgemm config dictionary and add the new entry for the operator to it, and expose the backend config with a function
+* Define/modify the lowering function of the custom backend that can fuse the pattern
+
+```python
+from torch.quantization.quantize_fx import prepare_fx
+from torch.quantization.quantize_fx import get_backend_config_dict, QuantizedOperatorType, update_operator_config
+import torch.fx
+
+
+# Let's say we want to modify the backend_config of fbgemm and add new quantized operator
+custom_backend_config_dict = get_backend_config_dict("fbgemm")
+bmm_config = {
+    "type": QuantizedOperatorType.NEED_OBSERVER_FOR_BOTH_INPUTS_AND_OUTPUTS,
+}
+update_operator_config(custom_backend_config_dict, torch.bmm, bmm_config)
```

Comment:
why `update_operator_config` isn't it just a dict assignment?

Reply:
this is a helper function to hide the complicated nesting in the configuration

Reply:
these are mostly for demo purposes, not final apis